lib/template/templates.c

```diff
@@ -28,11 +28,90 @@
 #include "template/escaping.h"
 #include "cfg.h"
 
+gboolean
+log_template_is_trivial(LogTemplate *self)
+{
+  return self->trivial;
+}
+
+const gchar *
+log_template_get_trivial_value(LogTemplate *self, LogMessage *msg, gssize *value_len)
+{
+  g_assert(self->trivial);
+
+  LogTemplateElem *e = (LogTemplateElem *) self->compiled_template->data;
+
+  switch (e->type)
+    {
+    case LTE_MACRO:
+      if (e->text_len > 0)
+        {
+          if (value_len)
+            *value_len = e->text_len;
+          return e->text;
+        }
+      else if (e->macro == M_MESSAGE)
+        return log_msg_get_value(msg, LM_V_MESSAGE, value_len);
+      else if (e->macro == M_HOST)
+        return log_msg_get_value(msg, LM_V_HOST, value_len);
+      g_assert_not_reached();
+    case LTE_VALUE:
+      return log_msg_get_value(msg, e->value_handle, value_len);
+    default:
+      g_assert_not_reached();
+    }
+}
+
+static gboolean
+_calculate_triviality(LogTemplate *self)
+{
+  /* if we need to escape, that's not trivial */
+  if (self->escape)
+    return FALSE;
+
+  /* no compiled template */
+  if (self->compiled_template == NULL)
+    return FALSE;
+
+  /* more than one element */
+  if (self->compiled_template->next != NULL)
+    return FALSE;
+
+  LogTemplateElem *e = (LogTemplateElem *) self->compiled_template->data;
+
+  switch (e->type)
+    {
+    case LTE_FUNC:
+      /* functions are never trivial */
+      return FALSE;
+    case LTE_MACRO:
+      /* Macros are trivial if they only contain a text but not a real
+       * macro.  Empty strings are represented this way.  */
```

Comment:
I don't fully understand this part. Could you explain for me what "only contain a text but not a real macro" means?

Reply:
A template is represented by a list of LogTemplateElem structs. Each one is a pair of a (literal text, dynamic reference).

Formatting works by first appending the literal text, then expanding the dynamic reference.

The reason this was combined was performance (half the items on the list than if we'd split the pairs), but there's a good chance this representation could be improved without to much performance consequences. E.g. having twice the elements but in a linear array rather than a linked list.

Reply:
Oh I see, thank you.